knoten/csm.py

```diff
@@ -18,192 +21,128 @@ def default(self, obj):
+        return dispatcher.dispatch(args[1].__class__)(*args, **kw)
+    wrapper.register = dispatcher.register
+    update_wrapper(wrapper, func)
+    return wrapper
+
+
+class Sensor(object):
+    def __init__(self, image):
+        isd = csmapi.Isd(image)
+        plugins = csmapi.Plugin.getList()
+        for plugin in plugins:
+            num_models = plugin.getNumModels()
+            for model_index in range(num_models):
+                model_name = plugin.getModelName(model_index)
+                if plugin.canModelBeConstructedFromISD(isd, model_name):
+                    self.model = plugin.constructModelFromISD(isd, model_name)
+
+    @classmethod
+    def request_isd(cls, image, url='http://pfeffer.wr.usgs.gov/api/1.0/pds/')
+        """
+        Instantiate a CSM sensor model by first requesting an ISD from a remote
+        service and writing the ISD to a json file adjacent to the passed image.
+
+        The image is opened and the method attempts to parse the label using the
+        PVL library. The label is then encoded and passed to the service.
+
+        Parameters
+        ----------
+        image : str
+                path to the image file
+
+        url : str
+              The URL of thge remote service used to generate an ISD.
+        """
+        label = pvl.dumps(image).decode()
+        image = json.dumps(label, cls=NumpyEncoder)
+        response = requests.post(url, data=image)
+        fname = os.path.splitext(image)[0] + '.json'
+        with open(fname, 'w') as f:
+            json.dump(response.json(), f)
+        
+        return cls(image)
+
+    @property
+    def semimajor_radii(self):
+        ellipsoid = csmapi.SettableEllipsoid.getEllipsoid(self.model)
+        return ellipsoid.getSemiMajorRadius()
+
+    @property
+    def semiminor_radii(self):
+        ellipsoid = csmapi.SettableEllipsoid.getEllipsoid(self.model)
+        return ellipsoid.getSemiMinorRadius()
+
+    @methodispatch
+    def intersect(self, height, image_pt):
+        if not isinstance(image_pt, csmapi.ImageCoord):
+            image_pt = csmapi.ImageCoord(*image_pt)
+        return self.model.imageToGround(image_pt, height)
+
+    @intersect.register(GeoDataset)
+    def _(self, dem, image_pt, tolerance=1, max_its=20):
+        if not isinstance(image_pt, csmapi.ImageCoord):
+            # Support a call where image_pt is in the form (x,y)
+            image_pt = csmapi.ImageCoord(*image_pt)
+
+        intersection = self.camera.imageToGround(image_pt, 0.0)
+        iterations = 0
+        while iterations < max_its:
+            height = 
+            next_intersection = camera.imageToGround(image_pt, height)
+            dist = max(abs(intersection.x - next_intersection.x),
+                    abs(intersection.y - next_intersection.y),
+                    abs(intersextion.z - next_intersection.z))
+
+            intersection = next_intersection
+            if dist < tolerance:
+                return intersection
```

Comment:
Returning None when we do not converge sounds hard to trace back down. We should either return what we have an a warning or raise and exception.